posthog/temporal/batch_exports/transformer.py

```diff
@@ -0,0 +1,233 @@
+                    data_gen = self.write(dumped + b"\n")
+            else:
+                # Orjson is very strict about invalid unicode. This slow path protects us
+                # against things we've observed in practice, like single surrogate codes, e.g.
+                # "\ud83d"
+                logger.exception("Failed to encode with orjson: %s", d)
+                cleaned_content = replace_broken_unicode(d)
+                data_gen = self.write(orjson.dumps(cleaned_content, default=str) + b"\n")
+        yield from data_gen
+
+    def _write_batch(self, record_batch: pa.RecordBatch) -> typing.Generator[bytes, None, None]:
+        """Write records to a temporary file as JSONL."""
+        for record_dict in record_batch.to_pylist():
+            if not record_dict:
+                continue
+
+            yield from self.write_dict(record_dict)
+
+
+class ParquetStreamTransformer(StreamTransformer):
+    def __init__(
+        self,
+        schema: pa.Schema,
+        compression: str | None = None,
+        compression_level: int | None = None,
+        include_inserted_at: bool = False,
+    ):
+        super().__init__(compression=compression, include_inserted_at=include_inserted_at)
+        self.schema = schema
+        self.compression_level = compression_level
+
+        # For Parquet, we need to handle schema and batching
+        self._parquet_writer = None
+        self._parquet_buffer = BytesIO()
+
+    @property
+    def parquet_writer(self) -> pq.ParquetWriter:
+        if self._parquet_writer is None:
+            self._parquet_writer = pq.ParquetWriter(
+                self._parquet_buffer,
+                schema=self.schema,
+                compression="none" if self.compression is None else self.compression,  # type: ignore
+                compression_level=self.compression_level,
+            )
+        return self._parquet_writer
+
+    def finalize(self):
+        """Ensure underlying Parquet writer is closed before flushing and closing temporary file."""
+        yield from super().finalize()
+
+        if self._parquet_writer is not None:
+            self._parquet_writer.close()
+            self._parquet_writer = None
+
+            # Get final data
+            self._parquet_buffer.seek(0)
+            final_data = self._parquet_buffer.read()
+            if final_data:
+                yield final_data
+
+            # Cleanup
+            self._parquet_buffer.close()
+            self._parquet_buffer = BytesIO()
+
+    def _write_batch(self, record_batch: pa.RecordBatch) -> typing.Generator[bytes, None, None]:
+        """Write records to a temporary file as Parquet."""
+
+        self.parquet_writer.write_batch(record_batch.select(self.parquet_writer.schema.names))
+        # Get current buffer content
+        self._parquet_buffer.seek(0)
+        data = self._parquet_buffer.read()
+
+        # Reset buffer
+        self._parquet_buffer.seek(0)
+        self._parquet_buffer.truncate()
```

Comment:
This feels wasteful: We are writing to a buffer, then copying the buffer data, and then deleting the buffer. Why not just write the data once, return that, and start a new buffer?

Reply:
Saves the copy.